### ISSUE TITLE
bridging between two emqx nodes

### ISSUE THREAD

**Issue by horowitzb** [emqx/emqx#2223]
#### Environment

- OS: Raspbian
- Erlang/OTP: 21.2
- EMQ: EMQX 3.0.1

#### Description

*any attempt to use emqx_ctl bridges causes an error*

whenever I try to bridge two nodes, I get an error. If I try to start the bridge as documentation says, this is the problem I get:
```
pi@pi03:~/Downloads/emq-relx/_rel/emqx $ ./bin/emqx_ctl bridges start emqx4@192.168.2.204 topic/#
bridges list                                    # List bridges
bridges start <Name>                            # Start a bridge
bridges stop <Name>                             # Stop a bridge
bridges forwards <Name>                         # Show a bridge forward topic
bridges add-forward <Name> <Topic>              # Add bridge forward topic
bridges del-forward <Name> <Topic>              # Delete bridge forward topic
bridges subscriptions <Name>                    # Show a bridge subscriptions topic
bridges add-subscription <Name> <Topic> <Qos>   # Add bridge subscriptions topic
bridges del-subscription <Name> <Topic>         # Delete bridge subscriptions topic
```
this suggests to me that that it's improper syntax, and I should use either `bridges start <Name>` or `bridges add-forward <Name> <Topic>` or `bridges add-subscription <Name> <Topic> <Qos>` to try and bridge the brokers. They all throw a similar error:

```
pi@pi03:~/Downloads/emq-relx/_rel/emqx $ ./bin/emqx_ctl bridges start emqx4@192.168.2.204
RPC to emqx3@192.168.2.203 error: {noproc,
                                   {gen_server,call,
                                    ['emqx_bridge_emqx4@192.168.2.204',
                                     start_bridge]}}
```

```
pi@pi03:~/Downloads/emq-relx/_rel/emqx $ ./bin/emqx_ctl bridges add-forward emqx4@192.168.2.204 topic/#
RPC to emqx3@192.168.2.203 error: {noproc,
                                   {gen_server,call,
                                    ['emqx_bridge_emqx4@192.168.2.204',
                                     {add_forward,<<"topic/#">>}]}}
```

```
pi@pi03:~/Downloads/emq-relx/_rel/emqx $ ./bin/emqx_ctl bridges add-subscription emqx4@192.168.2.204 topic/# 2
RPC to emqx3@192.168.2.203 error: {noproc,
                                   {gen_server,call,
                                    ['emqx_bridge_emqx4@192.168.2.204',
                                     {add_subscription,<<"topic/#">>,2}]}}
```

I already have node.name = emqx4@192.168.2.204 in the emqx.conf file for the other machine (and emqx3@192.168.2.203 in the config file for this machine)

I've already tried adding -name emqx4@192.168.2.204 to the vm.args file (and -name emqx3@192.168.2.203 for this machine)

And I know they should be able to connect, as clustering the two nodes together does work as intended.

What am I missing?

**Comment by horowitzb** [comment]
apparently when I enter  `./bin/emqx_ctl bridges start aws`, I get `start bridge successfully.`. the aws bridge seems to be defined in the emqx.conf file, so maybe I should make a new section in emqx.conf for every bridge I want to make?

**Comment by Gilbert-Wong** [comment]
For now, yes, we’ll improve this implementation in emqx 3.1 beta1

Gilbert Wong
On Feb 22, 2019, 11:28 PM +0800, horowitzb <notifications@github.com>, wrote:

apparently when I enter ./bin/emqx_ctl bridges start aws, I get start bridge successfully.. the aws bridge seems to be defined in the emqx.conf file, so maybe I should make a new section in emqx.conf for every bridge I want to make?

**Comment by horowitzb** [comment]
one further question: has the ability to both publish and subscribe to  a bridge been implemented yet? I'm able to subscribe through a bridge, but there doesn't seem to be an option to publish (that I can find). And if I try to create a bridge between both brokers, and subscribe to a topic on both, publishing to that topic from a client causes an infinite loop.

**Comment by jmpmscorp** [comment]
I'm at @horowitzb same point.

I want to bridge two emqx nodes. I have installed emqx edge v.3.1.beta1 on raspberry 3b and I want to bridge it to my central emqx broker installed on ubuntu server 16.04. On raspberry, emqx edge version, I have configured my 'emqx.conf' following this instructions:
https://developer.emqx.io/docs/tutorial/en/bridge/emqx_to_mosquitto.html

Bridge section on my 'emqx.conf' looks like:
```
##--------------------------------------------------------------------
## Test Bridge
##--------------------------------------------------------------------

bridge.test.address = 192.168.1.50:1883
bridge.test.start_type = auto
bridge.test.reconnect_interval = 30s
bridge.test.proto_ver = mqttv4
bridge.test.client_id = emqx_edge_test
bridge.test.clean_start = false
bridge.test.keepalive = 60s
bridge.test.mountpoint = test/
bridge.test.forwards = test1
bridge.test.subscription.1.topic = test/#
bridge.test.subscription.1.qos = 1
```

Opening dashboard of emqx central node, I can see raspberry node connected and, in subscriptions tab, I can see bridge subscription. Locally, publish messages works as expected. However, messages published in raspberry node are not forwarded to central node. I have tried with mosquitto_pub/sub and mqttfx clients but results are the same.

Furthermore, although I have configured keepalive as 60s, central dashboard shows 34560 as keepalive time.

Any new about new bridge implementation and configuration?

Regards.

**Comment by phanimahesh** [comment]
@jmpmscorp Bridging brokers carries the risk of looping messages infinitely. To avoid this, topics have to be chosen properly along with forwards and mountpoints.

#### Here is how your configuration is interpreted:

Your `forwards` is set to `test1`. This will mean messages published exactly on the topic `test` on emqx will be forwarded to `192.168.1.50:1883`. If you want multiple topics to be forwarded, separate them with a comma (`,`). Wildcards `+` and `#` are supported here with usual interpretation.

Messages published to the tree `test/#` on the remote broker will be available from emqx as the tree `test/test/#`.

If the behaviour you observe is different from this, let please file a new issue describing it.

**Comment by jmpmscorp** [comment]
@phanimahesh Thank you for reply.

I follow your instructions but, or I misunderstand something, or my bridge doesn't work as expected. I have tested all I can but nothing seems to work.

I have edited my emqx.conf on both, raspberry and central ubuntu, and allow to subscribe to # topics, so I must see any traffic between brokers.

I have connected two instances of MQTTfx, one to raspberry, one to emqx on ubuntu, and subscribe both to #. In mqttfx client connected to raspberry, I have published message through test1, test/test1, test/test/hello. As this client is subcribed to #, locally, every message is shown in subscription tab. However, anyone is show in central '#' subscription. 

On the other hand, I have published messages in central emqx in the same topics (test1, test/test1, test/test/hello). Anyone is shown on raspberry side. 

I don't know what is happened. I have another raspberry node with mosquitto broker instead emqx edge with same configuration and mosquitto->central emqx brdige works perfectly.

Another question... I understand that mountpoint is a preffix of forwarded message. If I configure, ie, forwards to test/#, and mounpoint raspberry/, I understand that forwarded topic will be raspberry/test/#, is correct?. And, does mountpoint work on inverse way?. If I would public in raspberry/test/# on central node, bridge node will translate topic to test/# (deleting raspberry/).
